Network/DNS/Decode/Internal.hs

```diff
@@ -239,19 +239,23 @@ getDomain' sep1 = do
             let offset = n * 256 + d
             mo <- pop offset
             case mo of
-                Nothing -> do
-                    inp <- getInput
-                    case runSGet (getDomain' sep1) (B.drop offset inp) of
-                      Left (DecodeError err) -> fail err
-                      Left err               -> fail $ show err
-                      Right o  -> push pos (fst o) >> return (fst o)
+                Nothing
+                  | offset `elem` pointerStack -> return "" -- pointer loop
+                  | otherwise                  -> do
+                      inp <- getInput
+                      let newStack = offset : pointerStack
+                          target = B.drop offset inp
```

Comment:
Should we check that `offset` is less than length of `inp`?  Or just rely on `Data.ByteString.drop` returning an empty string, from which we'll fail to parse a domain (no length/ptr field for the first label)?

As for loop detection we could instead limit the pointer recursion depth. A sensibly compressed name should never have a pointer directly to another pointer, it should then just point at that pointer's target. So the target of any pointer should be a `real` label. Since it takes at least two bytes to encode a label, and wire-form domain names are at most 255 bytes, with thus, at most every other label a non-pointer we can stop at 127 pointers:
```
->(l1.(->a2.(->a3.(...(->a127.)....))))
```
So it is probably cheaper to limit the recursion depth to 127 than to look for cycles i the visited offsets...
Of course cycle detection works too, but it feels needlessly costly (am I doing premature optimization?)

Reply:
Both are done!